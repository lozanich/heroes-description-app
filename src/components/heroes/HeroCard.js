import React from "react";
import { Link } from "react-router-dom";

export const HeroCard = ({ hero }) => {
  const {
    id,
    superhero,
    publisher,
    alter_ego,
    first_appearance,
    characters,
  } = hero;

  return (
    <>
      <div
        className="card animate__animated animate__jello"
        style={{ width: 18 + "rem", maxWidth: 500, minHeight: 840 }}
      >
        <img
          className="card-img-top"
          src={`./assets/heroes/${id}.jpg`}
          alt={superhero}
        />
        <div className="card-body">
          <h5 className="card-title">{superhero}</h5>
          <ul className="list-group list-group-flush">
            <li className="list-group-item">{publisher}</li>
            {alter_ego !== characters && (
              <li className="list-group-item">{characters}</li>
            )}
            {/* <li className="list-group-item">{alter_ego}</li> */}
            <li className="list-group-item">{first_appearance}</li>
          </ul>
          <Link to={`./hero/${id}`} className="btn btn-primary">
            Detalles
          </Link>
        </div>
      </div>
    </>
  );
};
