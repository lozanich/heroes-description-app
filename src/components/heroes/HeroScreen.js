import React, { useMemo } from "react";
import { Redirect, useParams } from "react-router-dom";
import { getHeroById } from "../../selectors/getHeroById";
// import batman from "../../assets/heroes/dc-batman.jpg"; // contenido estatico

const heroImages = require.context("../../assets/heroes", true);

export const HeroScreen = ({ history }) => {
  const { heroeId } = useParams();
  const hero = useMemo(() => getHeroById(heroeId), [heroeId]);
  // const hero = getHeroById(heroeId);
  console.log(heroeId);

  if (!hero) {
    return <Redirect to="/" />;
  }
  const { superhero, publisher, alter_ego, first_appearance, characters } =
    hero;

  const handleReturn = () => {
    // console.log(history);
    if (history.length <= 2) {
      // console.log("ifff history push");
      history.push("/");
    } else {
      history.goBack();
    }
  };

  return (
    <div className="row mt-5 animate__animated animate__backInLeft">
      <div className="col-4">
        <img
          className="img-thumbnail"
          // src={`../assets/heroes/${heroeId}.jpg`} // desde public assets
          // src={batman} // cuando tenemos un import
          src={heroImages(`./${heroeId}.jpg`)}
          alt={superhero}
        />
      </div>

      <div className="col-8">
        <h3>{superhero}</h3>

        <ul className="list-group list-group-flush">
          <li className="list-group-item">
            <span>Alter ego {alter_ego}</span>
          </li>
          <li className="list-group-item">
            <span>Publisher {publisher}</span>
          </li>
          <li className="list-group-item">
            <span>First Appearence {first_appearance}</span>
          </li>
        </ul>

        <h5>Characters</h5>
        <p> {characters}</p>

        <button onClick={handleReturn} className="btn btn-outline-primary">
          Regresar
        </button>
      </div>
    </div>
  );
};
