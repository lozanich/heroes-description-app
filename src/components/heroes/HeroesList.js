import React, { useMemo } from "react";
import { getHeroesByPublisher } from "../../selectors/getHeoresByPublisher";
import { HeroCard } from "./HeroCard";

export const HeroesList = ({ publisher }) => {
  const heroes = useMemo(() => getHeroesByPublisher(publisher), [publisher]);

  // const heroes = getHeroesByPublisher(publisher);
  return (
    <>
      <div className="card-columns">
        {heroes.map((hero) => {
          return <HeroCard key={hero.id} hero={hero}></HeroCard>;
        })}
      </div>
    </>
  );
};
