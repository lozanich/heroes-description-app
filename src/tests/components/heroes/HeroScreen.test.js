import React from "react";
import "@testing-library/jest-dom";
import { mount, shallow } from "enzyme";
import { MemoryRouter, Route, Router } from "react-router";
import { HeroScreen } from "../../../components/heroes/HeroScreen";

describe("Pruebas para el componente <HeroScreen />", () => {
  // http://localhost:3000/searchhero?q=iron
  const history = {
    length: 10,
    push: jest.fn(),
    goBack: jest.fn(),
    listen: jest.fn(),
    location: {},
    method: jest.fn(),
  };

  test("debe de mostrarse correctamente el componente", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/hero"]}>
        <HeroScreen history={history} />
      </MemoryRouter>
    );
    expect(wrapper.find("Redirect").exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  test("Debe de mostrar un hero si el parametro existe y se encuentra", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/hero/marvel-iron"]}>
        <Route path="/hero/:heroeId" component={HeroScreen}></Route>
      </MemoryRouter>
    );

    expect(wrapper.find(".row").exists()).toBe(true);
  });

  test("Debe de regresar a la pantalla anterior con PUSH", () => {
    const history = {
      length: 1,
      push: jest.fn(),
      goBack: jest.fn(),
      listen: jest.fn(),
      location: {},
      method: jest.fn(),
    };

    const wrapper = mount(
      <MemoryRouter initialEntries={["/hero/marvel-iron"]}>
        <Route
          path="/hero/:heroeId"
          component={(props) => <HeroScreen history={history} />}
        ></Route>
      </MemoryRouter>
    );

    wrapper.find("button").simulate("click");

    expect(history.push).toHaveBeenCalledWith("/");
    expect(history.goBack).not.toHaveBeenCalled();
  });

  test("Debe de llamar al go back para regresar a la pantalla anterior", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/hero/marvel-iron"]}>
        <Route
          path="/hero/:heroeId"
          component={(props) => <HeroScreen history={history} />}
        ></Route>
      </MemoryRouter>
    );

    wrapper.find("button").simulate("click");

    expect(history.goBack).toHaveBeenCalledWith();
    expect(history.push).not.toHaveBeenCalled();
  });

  test("Debe de llamar el Redirect si el hero no existe ", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/hero/marvel-iron8912828"]}>
        <Route
          path="/hero/:heroeId"
          component={(props) => <HeroScreen history={history} />}
        ></Route>
      </MemoryRouter>
    );

    console.log(wrapper.find("Rediect").exists());
    console.log(wrapper.html());

    expect(wrapper.text()).toBe("");
  });
});
