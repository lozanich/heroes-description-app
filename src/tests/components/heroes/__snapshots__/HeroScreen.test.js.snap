// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas para el componente <HeroScreen /> debe de mostrarse correctamente el componente 1`] = `""`;
